init/lib/start.go

```diff
@@ -17,24 +17,68 @@ package lib
 import (
 	"io/ioutil"
 	"os"
-	"os/exec"
-	"strconv"
 
 	"github.com/pkg/errors"
+	"gopkg.in/validator.v2"
+	"gopkg.in/yaml.v2"
+
+	"github.com/palantir/go-java-launcher/launchlib"
 )
 
-// StartCommand starts the given command, outputting to var/log/startup.log and writing the resulting process's PID to
-// var/run/service.pid.
-func StartCommand(cmd *exec.Cmd, outputFile *os.File) error {
-	cmd.Stdout = outputFile
-	cmd.Stderr = outputFile
-	if err := cmd.Start(); err != nil {
+func StartService(procCmds []launchlib.ProcCmd) error {
+	for _, procCmd := range procCmds {
+		if err := startCommand(procCmd); err != nil {
+			return errors.Wrap(err, "failed to start at least one process")
+		}
+		if err := writePid(procCmd.Name, procCmd.Cmd.Process.Pid); err != nil {
+			return errors.Wrap(err, "failed to record at least one pid")
+		}
+	}
+	return nil
+}
+
+func startCommand(procCmd launchlib.ProcCmd) (rErr error) {
+	procCmd.Cmd.Stdout = procCmd.Stdout
+	procCmd.Cmd.Stderr = procCmd.Stdout
+	if err := procCmd.Cmd.Start(); err != nil {
 		return errors.Wrap(err, "failed to start command")
 	}
+	return nil
+}
 
-	if err := ioutil.WriteFile(Pidfile, []byte(strconv.Itoa(cmd.Process.Pid)), 0644); err != nil {
+func writePid(name string, pid int) error {
```

Comment:
It feels quite gross to me that we write pids by:
1. have the name and pid
2. write the name and pid
3. read the name and pid from where we just wrote it
4. add a new name and pid to the ones we have just read
5. write both names and pids

Reply:
Actually, I started to edit `writePid` to be `writePids` and take a `map[string]int` and do it in batch, but since we can't atomically start a batch of commands I had to defer a function that would write all the pids (in case one of the later commands failed to start) so that we would still record which processes are now running. That feels to me grosser than what we already have. Maybe I'm wrong and deferring like that is considered less gross in the golang world? I'm not really sure what call to make here.

Reply:
I think you were right with the deferring method.  Deferring isn't seen in any particularly negative way and this use case seems similar to clean-up actions for which defer is recommended.